File: src/main/kotlin/dev/murad/shipping/item/container/TugRouteClientHandler.java
package dev.murad.shipping.item.container;

import com.mojang.datafixers.util.Pair;
import dev.murad.shipping.network.SetRouteTagPacket;
import dev.murad.shipping.network.TugRoutePacketHandler;
import dev.murad.shipping.util.TugRoute;
import dev.murad.shipping.util.TugRouteNode;
import net.minecraft.client.Minecraft;
import net.minecraft.client.gui.GuiGraphics;
import net.minecraft.client.gui.components.ObjectSelectionList;
import net.minecraft.client.gui.components.events.GuiEventListener;
import net.minecraft.network.chat.Component;
import org.jetbrains.annotations.NotNull;

import javax.annotation.Nullable;
import java.util.Optional;

public class TugRouteClientHandler {

    private TugList widget;
    private TugRoute route;
    private boolean isOffHand;
    private Minecraft minecraft;
    private TugRouteScreen screen;

    public TugRouteClientHandler(TugRouteScreen screen, Minecraft minecraft, TugRoute route, boolean isOffHand) {
        this.route = route;
        this.isOffHand = isOffHand;
        this.screen = screen;
        this.minecraft = minecraft;
    }



    public TugList initializeWidget(int width, int height, int y0, int y1, int itemHeight) {
        this.widget = new TugList(minecraft, width, height, y0, itemHeight);
        for (int i = 0; i < route.size(); i++) {
            this.widget.add(route.get(i), i);
        }

        return this.widget;
    }

    public void deleteSelected() {
        TugList.Entry selected = widget.getSelected();
        if (selected != null) {
            int index = selected.index;
            route.remove(index);
            this.widget.children().remove(index);
            this.widget.setSelected(null);
            markDirty();
        }
    }

    public void moveSelectedUp() {
        TugList.Entry selected = widget.getSelected();
        if (selected != null) {
            int index = selected.index;
            if (index > 0) {
                TugRouteNode node = route.remove(selected.index);
                this.widget.children().remove(index);
                route.add(index - 1, node);
                this.widget.children().add(index - 1, selected);
                markDirty();
            }
        }
    }

    public void moveSelectedDown() {
        TugList.Entry selected = widget.getSelected();
        if (selected != null) {
            int index = selected.index;
            if (index < this.route.size() - 1) {
                TugRouteNode node = route.remove(selected.index);
                this.widget.children().remove(index);
                route.add(index + 1, node);
                this.widget.children().add(index + 1, selected);
                markDirty();
            }
        }
    }

    public void renameSelected(@Nullable String name) {
        TugList.Entry selected = widget.getSelected();
        if (selected != null) {
            int index = selected.index;
            this.route.get(index).setName(name);
            markDirty();
        }
    }

    public Optional<Pair<Integer, TugRouteNode>> getSelected() {
        TugList.Entry selected = widget.getSelected();
        if (selected != null) {
            int index = selected.index;
            return Optional.of(new Pair<>(index, route.get(index)));
        }

        return Optional.empty();
    }

    public void render(GuiGraphics graphics, int mouseX, int mouseY, float partialTicks) {
        this.widget.render(graphics, mouseX, mouseY, partialTicks);
    }

    public final class TugList extends ObjectSelectionList<TugList.Entry> {
        public TugList(Minecraft minecraft, int width, int height, int y0, int itemHeight) {
            super(minecraft, width, height, y0, itemHeight);

            //TODO setRenderBackground(false);
            //TODO setRenderTopAndBottom(false);
        }

        @Override
        public @NotNull Optional<GuiEventListener> getChildAt(double p_212930_1_, double p_212930_3_) {
            return super.getChildAt(p_212930_1_, p_212930_3_);
        }

        public void add(TugRouteNode node, int index) {
            this.addEntry(new Entry(node, index));
        }

        @Override
        public int getRowWidth() {
            return screen.getXSize() - 40;
        }

        @Override
        protected int getScrollbarPosition() {
            return (this.width + getRowWidth()) / 2 + 5;
        }

        public class Entry extends ObjectSelectionList.Entry<Entry> {
            private TugRouteNode node;
            private int index;

            public Entry(TugRouteNode node, int index) {
                this.node = node;
                this.index = index;
            }

            @Override
            public void render(GuiGraphics graphics, int ind, int rowTop, int rowLeft, int width, int height, int mouseX, int mouseY, boolean hovered, float partialTicks) {
                String s = node.getDisplayName(index) + ": " + node.getDisplayCoords();

                graphics.blit(TugRouteScreen.GUI, rowLeft, rowTop, 0, hovered ? 216 : 236, width - 3, height);
                graphics.drawString(screen.getFont(), s, rowLeft + 3, rowTop + 4, 16777215);
            }

            public void setIndex(int index) {
                this.index = index;
            }

            @Override
            public boolean mouseClicked(double mouseX, double mouseY, int button) {
                if (button == 0) {
                    this.select();
                    return true;
                }
                return false;
            }

            private void select() {
                TugList.this.setSelected(this);
            }

            @Override
            public @NotNull Component getNarration() {
                // FIXME: ????
                return Component.literal("");
            }
        }
    }

    private void markDirty() {
        int i = 0;
        for (TugList.Entry entry : this.widget.children()) {
            entry.setIndex(i++);
        }

        SetRouteTagPacket setRouteTagPacket = new SetRouteTagPacket(route.hashCode(), isOffHand, route.toNBT());
        TugRoutePacketHandler.sendToServer(setRouteTagPacket);
    }
}

File: src/main/kotlin/dev/murad/shipping/item/container/TugRouteClientHandler.kt
package dev.murad.shipping.item.container

import com.mojang.datafixers.util.Pair
import dev.murad.shipping.network.SetRouteTagPacket
import dev.murad.shipping.network.TugRoutePacketHandler
import dev.murad.shipping.util.RouteNode
import dev.murad.shipping.util.TugRoute
import net.minecraft.client.Minecraft
import net.minecraft.client.gui.GuiGraphics
import net.minecraft.client.gui.components.ObjectSelectionList
import net.minecraft.network.chat.Component
import java.util.*

class TugRouteClientHandler(
    private val screen: TugRouteScreen,
    minecraft: Minecraft?,
    private val route: TugRoute,
    private val isOffHand: Boolean
) {
    private var widget: TugList? = null
    private val minecraft = minecraft!!


    fun initializeWidget(width: Int, height: Int, y0: Int, y1: Int, itemHeight: Int): TugList {
        this.widget = TugList(minecraft, width, height, y0, itemHeight)
        for (i in route.indices) {
            widget!!.add(route[i], i)
        }

        return this.widget!!
    }

    fun deleteSelected() {
        val selected = widget!!.selected
        if (selected != null) {
            val index = selected.getIndex()
            route.removeAt(index)
            widget!!.children().removeAt(index)
            widget!!.selected = null
            markDirty()
        }
    }

    fun moveSelectedUp() {
        val selected = widget!!.selected
        if (selected != null) {
            val index = selected.getIndex()
            if (index > 0) {
                val node = route.removeAt(selected.getIndex())
                widget!!.children().removeAt(index)
                route.add(index - 1, node)
                widget!!.children().add(index - 1, selected)
                markDirty()
            }
        }
    }

    fun moveSelectedDown() {
        val selected = widget!!.selected
        if (selected != null) {
            val index = selected.getIndex()
            if (index < route.size - 1) {
                val node = route.removeAt(selected.getIndex())
                widget!!.children().removeAt(index)
                route.add(index + 1, node)
                widget!!.children().add(index + 1, selected)
                markDirty()
            }
        }
    }

    fun renameSelected(name: String?) {
        val selected = widget!!.selected
        if (selected != null) {
            val index = selected.getIndex()
            route[index]!!.name = name
            markDirty()
        }
    }

    val selected: Optional<Pair<Int, RouteNode>>
        get() {
            val selected = widget!!.selected
            if (selected != null) {
                val index = selected.getIndex()
                return Optional.of(Pair(index, route[index]))
            }

            return Optional.empty()
        }

    fun render(graphics: GuiGraphics, mouseX: Int, mouseY: Int, partialTicks: Float) {
        widget!!.render(graphics, mouseX, mouseY, partialTicks)
    }

    inner class TugList(minecraft: Minecraft, width: Int, height: Int, y0: Int, itemHeight: Int) :
        ObjectSelectionList<TugList.Entry>(minecraft, width, height, y0, itemHeight) {


        fun add(node: RouteNode, index: Int) {
            this.addEntry(Entry(node, index))
        }

        override fun getRowWidth(): Int {
            return screen.xSize - 40
        }

        override fun getScrollbarPosition(): Int {
            return (this.width + rowWidth) / 2 + 5
        }

        inner class Entry(private val node: RouteNode, private var index: Int) : ObjectSelectionList.Entry<Entry>() {

            override fun render(
                graphics: GuiGraphics,
                ind: Int,
                rowTop: Int,
                rowLeft: Int,
                width: Int,
                height: Int,
                mouseX: Int,
                mouseY: Int,
                hovered: Boolean,
                partialTicks: Float
            ) {
                val s = node.getDisplayName(index) + ": " + node.getDisplayCoords()

                graphics.blit(
                    TugRouteScreen.GUI,
                    rowLeft,
                    rowTop,
                    0,
                    if (hovered) 216 else 236,
                    width - 3,
                    height
                )
                graphics.drawString(screen.getFont(), s, rowLeft + 3, rowTop + 4, 16777215)
            }

            fun setIndex(index: Int) {
                this.index = index
            }

            fun getIndex(): Int {
                return index
            }

            override fun mouseClicked(mouseX: Double, mouseY: Double, button: Int): Boolean {
                if (button == 0) {
                    this.select()
                    return true
                }
                return false
            }

            private fun select() {
                this@TugList.selected = this
            }

            override fun getNarration(): Component {
                // FIXME: ????
                return Component.literal("")
            }
        }
    }

    private fun markDirty() {
        var i = 0
        for (entry in widget!!.children()) {
            entry!!.setIndex(i++)
        }

        val setRouteTagPacket = SetRouteTagPacket(route.hashCode(), isOffHand, route.toNBT())
        TugRoutePacketHandler.sendToServer(setRouteTagPacket)
    }
}
